Reverse proxy support

vancluever opened the pull request:
Rebased, earlier text does not apply anymore... 

New commit text:

```
Reverse proxy support

This update adds full reverse proxy support, namely path-based reverse
proxy.

This is controlled by the -web.proxy-address flag, or the
NOMAD_PROXY_ADDRESS environment variable. Set this to the full path of
what the proxy looks like on the front end, ie:
https://example.com/nomad/.
```

vancluever commented:
Related: #156

jippi commented on frontend/index.html.ejs:
```diff
@@ -39,15 +14,51 @@
     var configScript = document.createElement('script');
-    configScript.setAttribute('src', '//' + NOMAD_ENDPOINT + '/config.js');
+    configScript.setAttribute('src', '//' + NOMAD_PRELOAD_ENDPOINT + '/config.js');
+    configScript.onload = function () {
```
i don't think we need this, if you move it into two <script> tags, then they will block each other, 0 race conditions :)

vancluever replied:
Yeah, this could be an artifact of my troubleshooting. I'll re-tool and see what happens.

I think I recall using relative paths to have no effect a while ago, at least when was trying to get the webpack-generated content to load. I will give this a shot - it would indeed be much simpler and I can move the various content points back to their respective places versus an `onload` parameter.

jippi commented on frontend/index.html.ejs:
```diff
@@ -39,15 +14,51 @@
-
     var configScript = document.createElement('script');
-    configScript.setAttribute('src', '//' + NOMAD_ENDPOINT + '/config.js');
+    configScript.setAttribute('src', '//' + NOMAD_PRELOAD_ENDPOINT + '/config.js');
```
couldn't we just always ask for `./config.js` without and path, then it would be whatever path you are on now + `config.json` which would satisfy the requirement in the Go server i think :)

jippi commented:
The changes seem to break development work-flow :) currently debugging

jippi commented:
Haven't gotten it to work, but leaving it for now :)

vancluever commented:
OK, I've added a WIP to this for now. I'll sort out the loading and dev issues and re-base.

vancluever commented:
OK, I think I found the root to a lot of my issues - since my initial URL (ie: `http://localhost/nomad`) did not have a trailing slash in it, `config.js` was loading off `http://localhost/nomad` regardless of whether I did `<script src="config.js" ...>` or `<script src="./config.js" ...>`. :angry: 

I'll just add the redirects in my proxy to make sure this doesn't happen.

Now that this is sorted, using relative paths everywhere should work for all the things - so I'm wondering how much we need the proxy address setting. Maybe in the javascript for things like loading websocket proxy addresses and detecting the base path for react routes, but can we get away with removing the "/" root path from the webpack config and have the main CSS and JS content get loaded from a relative path too? We might be able to clean up a good deal of the `appendChild` stuff that way. I think exposing config values is still valuable though.

Also, we can probably move `/config.js` back to a static route that's tracked by New Relic again.

vancluever commented:
I think this is now in a state where it should be acceptable - `config.js` loads off a single path cleanly, and also I've removed the `/` prefix from the webpack config so that the compiled CSS and JS files load off of relative paths from the base as well.

I've rebased this as the history was getting a bit messy, have a look and let me know what you think :) Dev mode should work now as well.

jippi commented:
@vancluever I'll review it today and merge it if everything works nicely :)

jippi commented on backend/main.go:
```diff
@@ -231,17 +242,33 @@ func main() {
+		if cfg.ProxyAddress != "" {
+			endpointURL, err = url.Parse(cfg.ProxyAddress)
+			if err != nil {
+				logger.Infof("Error parsing supplied web.proxy-address of %s: %s", cfg.ProxyAddress, err)
```
the check should be moved to the configuration parser / validation, so we avoid this as a runtime error :)

vancluever replied:
Hmm, so the only issue I see with that is the error handling here is not necessarily validation, but trapping errors that come from `url.Parse`. So if we wanted to validate this for the construction of `NOMAD_ENDPOINT`, we would have to execute this twice, and drop one of the results on the floor. It seems to me that putting this in validation we would have to either build `NOMAD_ADDR` there or test-parse it, and drop the result on the floor. The former seems like it would be polluting the `Config` struct with environment, and the latter seems like a duplication of work. Thoughts?

vancluever replied:
One thing I'm thinking of doing is just skipping the URL splitting mainly because most things in the code that depend on the url either 1) are using the currently loaded scheme anyway (i.e. in `sagas/event.js` and the React `main.js`) or have been removed anyway in favour of relative links. So I think we can just get away with getting rid of this parsing together (I was probably being a bit overzealous) and just expect the proxy URL to just be something like `example.com/nomad` without the scheme supplied.

I'll rebase this with the updated code tomorrow!

jippi replied:
@vancluever alright, let me know ! :)

vancluever commented:
@jippi just updated it, removed all the URL parsing logic and updated docs as to not imply the use of protocol scheme when supplying the proxy information. Smoke tested and it all looks good. Let me know when you have a chance to review!

PS: I also added `backend/backend` to `.gitignore` so that git didn't load in the `backend/backend` artifact that got created when `go build` was run on the code (part of my current edit/save/check workflow).

jippi commented:
@vancluever Seem to be missing flag parsing :) 

```diff
diff --git backend/main.go backend/main.go
index 9b59493..31f49f0 100644
--- backend/main.go
+++ backend/main.go
@@ -164,6 +164,10 @@ func (c *Config) Parse() {
        if *flagNewRelicLicense != "" {
                c.NewRelicLicense = *flagNewRelicLicense
        }
+
+       if *flagProxyAddress != "" {
+               c.ProxyAddress = *flagProxyAddress
+       }
 }

```

jippi commented:
@vancluever other than the missing flag parsing, i think its good to ship :) haven't test it in great length, as I don't have the infrastructure in place to do so :)

vancluever commented:
Hey @jippi - all done - added and good to go!

I've been testing this pretty extensively during development, so I think it should be okay :) also did a quick test now to make sure the option was being accepted vs the env var.

jippi commented:
I'm going to try and deploy it once the build is done, and if it works, I'll merge it ! :)

vancluever commented:
Thanks again for all the help in this @jippi!

jippi commented:
Thanks @vancluever ! https://github.com/jippi/hashi-ui/releases/tag/v0.10.0 should be on dockerhub in 10min time :)
